pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,484 @@
+		return
+	}
+
+	cachedPodIP, exists := c.getPodNameIPCache(podName)
+	if !exists && podIP == "" {
+		return
+	}
+
+	if exists {
+		if podIP == "" {
+			klog.InfoS("cacheIP", "ip", cachedPodIP, "pod", podName.String())
+			deletePodFlow(podName, cachedPodIP)
+			return
+		} else if cachedPodIP != podIP {
+			deletePodFlow(podName, cachedPodIP)
+			createOrUpdatePodFlow(podName, podIP)
+			return
+		}
+	}
+
+	needCreateOrUpdate := c.checkPod(pod)
+	if !needCreateOrUpdate {
+		return
+	}
+
+	createOrUpdatePodFlow(podName, podIP)
+}
+
+func (c *MCPodRouteController) Run(stopCh <-chan struct{}) {
+	defer c.gwQueue.ShutDown()
+	defer c.podQueue.ShutDown()
+
+	klog.InfoS("Starting controller", "controller", podRouteControllerName)
+	defer klog.InfoS("Shutting down controller", "controller", podRouteControllerName)
+	if !cache.WaitForNamedCacheSync(podRouteControllerName, stopCh, c.gwInformer.HasSynced) {
+		return
+	}
+	// Run a single routine to handle Gateway events.
+	go wait.Until(c.gatewayWorker, time.Second, stopCh)
+	<-stopCh
+}
+
+func (c *MCPodRouteController) gatewayWorker() {
+	for c.processGatewayNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processGatewayNextWorkItem() bool {
+	key, quit := c.gwQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gwQueue.Done(key)
+
+	if k, ok := key.(string); !ok {
+		c.gwQueue.Forget(k)
+		klog.InfoS("Expected string in work queue but got", "object", k)
+		return true
+	} else if err := c.syncGateway(); err == nil {
+		c.gwQueue.Forget(key)
+	} else {
+		c.gwQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing Gateway, requeuing", "key", key)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncGateway() error {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+		return err
+	}
+
+	c.podWorkersStartedMutex.Lock()
+	defer c.podWorkersStartedMutex.Unlock()
+
+	isMyselfGateway := activeGW != nil && c.nodeConfig.Name == activeGW.Name
+	// Stop Pod flow controller and clean up all installed Multi-cluster Pod flows,
+	// if the Node was a Gateway before.
+	if !isMyselfGateway {
+		if c.podWorkersStarted {
+			klog.InfoS("Shutting down Multi-cluster PodFlowController")
+			close(c.podWorkerStopCh)
+			c.podWorkerStopCh = nil
+			c.podInformer = nil
+			c.podLister = nil
+			c.podWorkersStarted = false
+		}
+	}
+
+	if c.getPodIPsCacheSize() > 0 && (!isMyselfGateway || !c.podWorkersStarted) {
+		err := c.ofClient.UninstallMulticlusterPodFlows("")
+		if err != nil {
+			return err
+		}
+		c.podIPsCache = map[string]string{}
+		return nil
+	}
+
+	if isMyselfGateway {
+		if !c.podWorkersStarted {
+			klog.InfoS("Starting Multi-cluster PodFlowController")
+			c.podWorkerStopCh = make(chan struct{})
+			c.createPodInformer()
+			go c.podInformer.Run(c.podWorkerStopCh)
+			if !cache.WaitForNamedCacheSync(podRouteControllerName, c.podWorkerStopCh, c.podInformer.HasSynced) {
+				close(c.podWorkerStopCh)
```

Comment:
`WaitForNamedCacheSync` will only return false when `podWorkerStopCh` is closed. While there is no way to close that channel from current code, it can't close the channel the second time which would lead to panic.

Reply:
removed the line